### PR TITLE
feat(review-workflows): apply license restrictions

### PR DESCRIPTION
**Pull request by nathan-pichon**
### What does it do?

Adding EE licenses limits options to multiple workflows.
Depending on the plan, the number of available workflows and a number of stages per workflow can be restricted.
Also adding a hard limit to mitigate potential issues (e.g. infinite loop on workflow creation)

### Why is it needed?

As Review Workflows is an EE feature, we need to be able to limit how users can use the feature on Cloud plans.
Self-hosted EE is not impacted by those limitations.

### How to test it?

- Create workflows and try to bypass the limit of workflows available for your license.
- Create stages on a workflow and try to bypass the limit of stagesPerWorkflow available for your license.

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

### CODE REVIEW

File: packages/core/admin/ee/server/services/review-workflows/workflows/index.js [hunk 1]
```diff
@@ -19,8 +27,21 @@ const processFilters = ({ strapi }, filters = {}) => {
...
   const workflowsContentTypes = workflowsContentTypesFactory({ strapi });
+  const limits = {
+    workflows: MAX_WORKFLOWS,
+    stagesPerWorkflow: MAX_STAGES_PER_WORKFLOW,
+  };
 
   return {
+    register({ workflows, stagesPerWorkflow }) {
+      if (!Object.isFrozen(limits)) {
+        limits.workflows = clampMaxWorkflows(workflows || limits.workflows);
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/workflows/index.js:38]: super minuscule thing, is it really necessary to have the clamp utility? We will only be usin git in the workflows/stages register and I feel it's more expressive to have 
`limits.workflows = clamp(1, MAX_WORKFLOWS, workflows ?? MAX_WORKFLOWS);`

Comment (Marc-Roig) [reply packages/core/admin/ee/server/services/review-workflows/workflows/index.js:38]: that or:
```js
    register({ workflows = MAX_WORKFLOWS, stagesPerWorkflow = MAX_STAGES_PER_WORKFLOW }) {
      if (!Object.isFrozen(limits)) {
        limits.workflows = clamp(1, MAX_WORKFLOWS, workflows);
```

Comment (nathan-pichon) [reply packages/core/admin/ee/server/services/review-workflows/workflows/index.js:38]: As we do not use Typescript, I would rather prefer to use `||` to avoid any `NaN` because we pass a null value. 😅 

The clamp function is important as it will ensure that the value is between 1 and MAX_WORKFLOWS.

[the remark this answers is not in the document]
Comment (nathan-pichon) [reply packages/core/admin/ee/server/services/review-workflows/workflows/index.js:38]: I don't think it makes sense to have 0 workflows as the limit, as it would lock the feature with our default workflow.
If so, we might "just" want to disable review workflow 🤔

File: packages/core/admin/ee/server/services/review-workflows/stages.js [hunk 2]
```diff
@@ -30,6 +31,17 @@ module.exports = ({ strapi }) => {
...
+        if (!workflows[stage.workflowId]) {
+          workflows[stage.workflowId] = [];
+        }
+        workflows[stage.workflowId].push(stage);
+        return workflows;
+      };
+      // As we can create several stages, we need to make sure that we don't exceed the licence threshold
+      Object.entries(stagesList.reduce(groupByWorkflow, {})).forEach((workflowStages) => {
+        workflowsValidationService.validateWorkflowStages(workflowStages);
+      });
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/stages.js:44]: arent all the stages from the list using the same workflow always?

Comment (nathan-pichon) [reply packages/core/admin/ee/server/services/review-workflows/stages.js:44]: In this function, nothing is preventing developer from adding stages to any workflow 🤔

File: packages/core/admin/ee/server/services/review-workflows/workflows/content-types.js [hunk 3]
```diff
@@ -28,12 +28,17 @@ module.exports = ({ strapi }) => {
...
-    async migrate({ srcContentTypes = [], destContentTypes, stageId }) {
+    async migrate({ srcContentTypes = [], destContentTypes, stageId }, force = false) {
       // Workflows service is using this content-types service, to avoid an infinite loop, we need to get the service in the method
       const workflowsService = getService('workflows', { strapi });
+      const workflowsValidationService = getService('review-workflows-validation', { strapi });
       const { created, deleted } = diffContentTypes(srcContentTypes, destContentTypes);
 
+      if (!force) {
+        await workflowsValidationService.validateWorkflowCount();
+      }
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/workflows/content-types.js:41]: Is it necessary to validate the workflow count at this point? Imo the workflow service itself must know about its validations, but its not necessary for the content-type migrations itself to do so.

By looking at the code it seems all workflow service methods using this migrate are already validating the workflow count.

Comment (nathan-pichon) [reply packages/core/admin/ee/server/services/review-workflows/workflows/content-types.js:41]: This is true, but any part of the app can load this service and use the migrate function outside of the workflow service

File: packages/core/admin/ee/server/services/review-workflows/stages.js (hunk 2 above)
Comment (Marc-Roig) [reply packages/core/admin/ee/server/services/review-workflows/stages.js:44]: It would seem to me as a missuse of the method, but still !

This is my shining opportunity to propose a fp approach ✨ 

```js
const groupByWorkflow = (stage) => stage.workflowId;

pipe(
  groupBy(groupByWorkflow),
  forEach(workflowsValidationService.validateWorkflowStages)
)(stagesList);
```

Still, I think this only will validate the amount of stages the user is creating, but not how many stages the workflow has. So I don't see much value in it.

**Review by Marc-Roig — APPROVED** [review]
Only minor comments :) Looks good to me!

File: packages/core/admin/ee/server/services/review-workflows/workflows/index.js [hunk 4]
```diff
@@ -49,30 +50,32 @@ module.exports = ({ strapi }) => {
...
           await workflowsContentTypes.migrate({
             destContentTypes: opts.data.contentTypes,
-            stageId: stageIds[0],
+            stageId: stages[0].id,
           });
         }
 
-        // Create Workflow
-        return strapi.entityService.create(WORKFLOW_MODEL_UID, createOpts);
+        return { ...workflow, stages };
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/workflows/index.js:78]: I don't see why this change is necessary if you are already validating the worklfow number and number of stages before anything else.

File: packages/core/admin/ee/server/services/review-workflows/workflows/index.js [hunk 5]
```diff
@@ -49,30 +50,32 @@ module.exports = ({ strapi }) => {
...
+      await workflowsValidationService.validateWorkflowCount(1);
 
       return strapi.db.transaction(async () => {
-        // Create stages
-        const stageIds = await getService('stages', { strapi })
-          .replaceStages([], opts.data.stages)
-          .then((stages) => stages.map((stage) => stage.id));
+        // Create Workflow
+        const workflow = await strapi.entityService.create(WORKFLOW_MODEL_UID, {
+          select: '*',
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/workflows/index.js:59]: I think all fields are selected by default if you don't provide this parameter.

File: packages/core/admin/ee/server/services/review-workflows/stages.js [hunk 6]
```diff
@@ -77,20 +85,13 @@ module.exports = ({ strapi }) => {
...
       return strapi.db.transaction(async ({ trx }) => {
-        // Create the new stages
-        const createdStages = await this.createMany(created, { fields: ['id'] });
-        // Put all the newly created stages ids
-        const createdStagesIds = map('id', createdStages);
-
-        // Update the workflow stages
-        await mapAsync(updated, (stage) => this.update(stage.id, stage));
-
+        // ⚠️ Delete first as we can have a limit on stages
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/stages.js:88]: I believe this comment can be deleted

File: packages/core/admin/ee/server/services/review-workflows/stages.js [hunk 7]
```diff
@@ -77,20 +85,13 @@ module.exports = ({ strapi }) => {
...
         // Delete the stages that are not in the new stages list
         await mapAsync(deleted, async (stage) => {
           // Find the nearest stage in the workflow and newly created stages
           // that is not deleted, prioritizing the previous stages
           const nearestStage = findNearestMatchingStage(
-            [...srcStages, ...createdStages],
+            srcStages,
```
Comment (Marc-Roig) [inline packages/core/admin/ee/server/services/review-workflows/stages.js:94]: I think we should keep the original order here. the source stages also need the created stages at the end. If you delete all stages and add a new one, the computed `nearestStage` will be undefined, and entities stage will be deleted.

File: packages/core/admin/ee/server/services/review-workflows/stages.js (hunk 6 above)
Comment (nathan-pichon) [reply packages/core/admin/ee/server/services/review-workflows/stages.js:88]: Why? 🤔

File: packages/core/admin/ee/server/services/review-workflows/stages.js (hunk 7 above)
Comment (nathan-pichon) [reply packages/core/admin/ee/server/services/review-workflows/stages.js:94]: We can't have the created stages here as the creation is done afterward.

Comment (Marc-Roig) [reply packages/core/admin/ee/server/services/review-workflows/stages.js:94]: but we should because of that point

File: packages/core/admin/ee/server/services/review-workflows/stages.js (hunk 6 above)
Comment (Marc-Roig) [reply packages/core/admin/ee/server/services/review-workflows/stages.js:88]: there is no validation done here

File: packages/core/admin/ee/server/services/review-workflows/stages.js (hunk 7 above)
Comment (nathan-pichon) [reply packages/core/admin/ee/server/services/review-workflows/stages.js:94]: Oh but as I removed the validation, I may not need to reorder things here too. Will change this
